### ISSUE TITLE
GKE preemptible node recreates, causes non-unique pod name

### ISSUE THREAD

**Issue by kim0** [kubernetes/kubernetes#64067]
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
/kind bug
> /kind feature

**What happened**:
I am using GKE 1.9.7. Currently using one or two preemptible nodes for cost saving. My understanding is that the preemptible node is destroyed/recreated roughly every 24 hours. This causes a few minutes of pod disruption but I dont mind. I noticed after the node was recreated, my deployment created a pod having the same name as a pod a couple of days ago. This caused errors in my monitoring since this type of thing should not happen. I'm pasting relevant logs below from stackdriver logging

Some notes

* The repeated names occur at exactly the time at which the preemptible node should be replaced! I'm not sure whether precisely whether this is before/during/or-after the node is fully replaced
* The application emits a "special startup message" which can be seen below in logs. This (among other things) tells me the app/pod was restarted, but ended up having repeated pod name. 

Let me know how to collect more debug info to verify whether what I'm see'ing really did happen

OLD pod startup logs
```
{
 insertId:  "1acjn9bg2t9b9b8"  
 labels: {…}  
 logName:  "projects/myprojectfoobar/logs/poddy"  
 receiveTimestamp:  "2018-05-16T15:39:10.359342818Z"  
 resource: {
  labels: {
   cluster_name:  "hash-eu"    
   container_name:  "poddy"    
   instance_id:  "6165026643360477450"    
   namespace_id:  "default"    
   pod_id:  "poddy-65779bcd64-6wd2b"    
   project_id:  "myprojectfoobar"    
   zone:  "europe-west1-b"    
  }
  type:  "container"   
 }
 severity:  "INFO"  
 textPayload:  " * MODE        special startup message "  
 timestamp:  "2018-05-16T15:38:34Z"  
}
```

Repeated POD startup logs

```


{
 insertId:  "dm1vbdfa1ihd0"  
 labels: {
  compute.googleapis.com/resource_name:  "fluentd-gcp-v2.0.17-4h9j6"   
  container.googleapis.com/namespace_name:  "default"   
  container.googleapis.com/pod_name:  "poddy-65779bcd64-6wd2b"   
  container.googleapis.com/stream:  "stdout"   
 }
 logName:  "projects/myprojectfoobar/logs/poddy"  
 receiveTimestamp:  "2018-05-17T15:42:00.406746199Z"  
 resource: {
  labels: {
   cluster_name:  "hash-eu"    
   container_name:  "poddy"    
   instance_id:  "8378159635278863461"    
   namespace_id:  "default"    
   pod_id:  "poddy-65779bcd64-6wd2b"    
   project_id:  "myprojectfoobar"    
   zone:  "europe-west1-b"    
  }
  type:  "container"   
 }
 severity:  "INFO"  
 textPayload:  " * MODE        special startup message "  
 timestamp:  "2018-05-17T15:41:53Z"  
}
```

**What you expected to happen**: PODs to have globally and temporally unique names!

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): 1.9.7
- Cloud provider or hardware configuration: GKE
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

**Comment by kim0** [comment]
Per twitter yesterday @janetkuo .. Here's the bug report. Thanks

**Comment by janetkuo** [comment]
Thanks @kim0 for reporting this. 

Deployment only controls the prefix of a pod's name (Deployment name + pod template hash). The suffix is random string generated by API server. The chances of generating the same pod name for a Deployment should be pretty low. 

However, it's not impossible, as Kubernetes only guarantees that no resource of the same kind (e.g. `Pod`) and in the same namespace (e.g. `default`) could have the same name at the same time. 

@kim0 is the "special startup message" generated by your app/pod? Why does it think the pod is restarted, does it do something like looking for pod names in the logs?

Also, Does this Deployment have large `.spec.replicas`? It could increase the chance of having the same suffix.

**Comment by kim0** [comment]
Yes, the startup message is generated by the app/pod, it is emitted when the pod is restarted. And yes, the I'm recording some billing data together with pod names in a DB. I was previously assuming that max(billing_data) was the value I wanted. But if a pod can recreate (making billing_data=0) with the same name later, it's not. Anyway, I changed how I calculate the billing info, so I'm good. And I think you're saying the chances of pod name collision as long as they're not at the same time is non-zero, so you're probably saying this is not really a bug.

Since I've seen this multiple times over the course of one month, I think something is making this event happen with a much higher frequency that it should. Perhaps bec I'm only working on a single node, which when preempted the k8s masters are unable to schedule any pods anywhere or so. But I'm not sure if this qualifies as a bug since if it can be allowed to happen rarely, it might as well be allowed to happen more frequently :) In fact I'm glad I saw it happen and corrected how I calculate billing, instead of assuming pod suffixes were unique over all time. Thanks for helping!
